apps/admin/frontend/src/components/save_file_modal.tsx

```diff
@@ -0,0 +1,261 @@
+  isSystemAdministratorAuth,
+  usbstick,
+} from '@votingworks/utils';
+import type { ExportDataError } from '@votingworks/admin-backend';
+
+import {
+  Button,
+  Modal,
+  H1,
+  P,
+  useExternalStateChangeListener,
+  Font,
+} from '@votingworks/ui';
+
+import { AppContext } from '../contexts/app_context';
+import { Loading } from './loading';
+
+export const UsbImage = styled.img`
+  margin-right: auto;
+  margin-left: auto;
+  height: 200px;
+`;
+
+interface SaveAsButtonProps {
+  onSave: (location: string) => void;
+  options?: KioskBrowser.SaveDialogOptions;
+}
+
+export function SaveAsButton({
+  onSave,
+  options,
+}: SaveAsButtonProps): JSX.Element {
+  async function useSaveDialog() {
+    assert(window.kiosk);
+    const { filePath } = await window.kiosk.showSaveDialog(options);
+    if (filePath) {
+      onSave(filePath);
+    }
+  }
+  return <Button onPress={useSaveDialog}>Save As…</Button>;
+}
+
+export type SaveFileError =
+  | ExportDataError
+  | {
+      type: 'api-error';
+      message: string;
+    };
+
+export type SaveFileResult = Result<string[], SaveFileError>;
+
+export interface SaveFileModalProps {
+  onSave: (location: string) => Promise<SaveFileResult>;
+  onClose: () => void;
+  /**
+   * Title case name for the type of file being saved, e.g. "Batch Results".
+   */
+  fileTypeTitle: string;
+  /**
+   * Lowercase name for the type of file being saved, e.g. "batch results".
+   */
+  fileType: string;
+  /**
+   * Relative to the root of the USB drive.
+   */
+  defaultRelativePath: string;
+}
+
+type SaveFileModalState = 'error' | 'saving' | 'done' | 'init';
+
+export function SaveFileModal({
+  onSave,
+  onClose,
+  fileTypeTitle,
+  fileType,
+  defaultRelativePath,
+}: SaveFileModalProps): JSX.Element {
+  const { usbDrive, auth } = useContext(AppContext);
+  assert(isElectionManagerAuth(auth) || isSystemAdministratorAuth(auth));
+
+  const [currentState, setCurrentState] = useState<SaveFileModalState>('init');
+  const [errorMessage, setErrorMessage] = useState('');
+  const [usbDrivePath, setUsbDrivePath] = useState<string>();
+
+  useExternalStateChangeListener(usbDrive.status, async (newStatus) => {
+    if (newStatus === 'mounted') {
+      const usbDriveInfo = await usbstick.getInfo();
+      assert(usbDriveInfo);
+      setUsbDrivePath(usbDriveInfo.mountPoint);
+    }
+  });
+
+  async function saveFile(location: string) {
+    setCurrentState('saving');
+    const result = await onSave(location);
+    if (result.isOk()) {
+      setCurrentState('done');
+    } else {
+      setCurrentState('error');
+      const error = result.err();
+      switch (error.type) {
+        case 'permission-denied':
+          setErrorMessage('Permission denied.');
+          break;
+        case 'file-system-error':
+          setErrorMessage('There may be an issue with the USB drive.');
+          break;
+        case 'missing-usb-drive':
+        case 'relative-file-path':
+        case 'api-error':
```

Comment:
All of these errors are unexpected and would be the equivalent of 500 errors, so not handling specifically.